File: solutions/day6.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const lab = raw.split("\n").map((row) => row.split(""));
const directions = ["up", "right", "down", "left"];
const directionsMap = new Map<string, [number, number]>([
  ["up", [0, -1]],
  ["right", [1, 0]],
  ["down", [0, 1]],
  ["left", [-1, 0]],
]);

function locateStart(map: string[][]): [number, number] {
  for (let i = 0; i < map.length; i++) {
    for (let j = 0; j < map[i].length; j++) {
      if (map[i][j] === "^") return [j, i];
    }
  }
  return [-1, -1];
}

// function displayLab(map: string[][]): void {
//   console.log(map.map((row) => row.join("")).join("\n"));
//   console.log();
// }

function iterate(
  map: string[][],
  [x, y]: [number, number],
  direction: string,
  isPart1: boolean,
  statesSet: Set<string>
): {
  conditions: { leftArea: boolean; enteredLoop: boolean };
  coords: [number, number];
  direction: string;
} {
  let [dx, dy] = directionsMap.get(direction)!;
  let newX = x + dx;
  let newY = y + dy;
  if (
    newX < 0 ||
    newX > map[0].length - 1 ||
    newY < 0 ||
    newY > map.length - 1
  ) {
    if (isPart1) map[y][x] = "X";
    return {
      conditions: { leftArea: true, enteredLoop: false },
      coords: [newX, newY],
      direction,
    };
  }

  const rotate = (direction: string) => {
    if (map[newY][newX] !== "#") return direction;
    const newDirection =
      directions[(directions.indexOf(direction) + 1) % directions.length];
    [dx, dy] = directionsMap.get(newDirection)!;
    newX = x + dx;
    newY = y + dy;
    return rotate(newDirection);
  };
  const newDirection = rotate(direction);

  const temp = map[y][x];
  map[y][x] = isPart1 ? "X" : map[newY][newX];
  map[newY][newX] = temp;

  if (statesSet.has(JSON.stringify([x, y, direction]))) {
    return {
      conditions: { leftArea: false, enteredLoop: true },
      coords: [newX, newY],
      direction: newDirection,
    };
  }
  statesSet.add(JSON.stringify([x, y, direction]));

  return {
    conditions: { leftArea: false, enteredLoop: false },
    coords: [newX, newY],
    direction: newDirection,
  };
}

// Part 1
const newLab: string[][] = JSON.parse(JSON.stringify(lab));
let [x, y] = locateStart(newLab);
let direction = directions[0];
let leftArea = false;
const statesSet = new Set<string>();
while (!leftArea) {
  const nextState = iterate(newLab, [x, y], direction, true, statesSet);
  leftArea = nextState.conditions.leftArea;
  [x, y] = nextState.coords;
  direction = nextState.direction;
}

const uniqueLocations = newLab.reduce(
  (acc, row) => acc + row.filter((position) => position === "X").length,
  0
);
console.log("Part 1:", uniqueLocations);

// Part 2
const loops = lab.reduce<number>(
  (acc, row, i) =>
    acc +
    row.reduce<number>((acc, _, j) => {
      const statesSet = new Set<string>();
      const newLab: string[][] = JSON.parse(JSON.stringify(lab));
      newLab[i][j] = "#";
      let [x, y] = locateStart(newLab);
      let direction = directions[0];
      let leftArea = false;
      let enteredLoop = false;
      while (!leftArea) {
        if (enteredLoop) {
          acc++;
          break;
        }
        const nextState = iterate(newLab, [x, y], direction, false, statesSet);
        leftArea = nextState.conditions.leftArea;
        enteredLoop = nextState.conditions.enteredLoop;
        [x, y] = nextState.coords;
        direction = nextState.direction;
      }
      return acc;
    }, 0),
  0
);
console.log("Part 2:", loops);

File: solutions/python/day6.py
with open("./input.txt", "r") as txt:
  raw: str = txt.read().rstrip()
  txt.close()

lab = [list(i) for i in raw.split('\n')]

directions = [
    'up',
    'right',
    'down',
    'left'
]

def locateRobot(map: list[list[str]]) -> tuple[int, int]:
    for i in range(len(map)):
        for j in range(len(map[i])):
            if map[i][j] == '^':
                return (j, i)
    # return (-1, -1)

def checkLab(map: list[list[str]]) -> None:
    for i in range(len(map)):
        for j in range(len(map[i])):
            print(map[i][j], end='')
        print()
    print()

def checkObstructions(map: list[list[str]], x: int, y: int, direction: str, isPart1: bool) -> tuple[bool, int, int, str]:
    dxdy = {
        'up': (0, -1),
        'right': (1, 0),
        'down': (0, 1),
        'left': (-1, 0)
    }
    dx, dy = dxdy[direction]
    newX, newY = x + dx, y + dy
    if newX < 0 or newX > len(map[0]) - 1 or newY < 0 or newY > len(map) - 1:
        if isPart1: map[y][x] = 'X'
        return (True, newX, newY, direction)
    
    while map[newY][newX] == '#':
        direction = directions[(directions.index(direction) + 1) % len(directions)]
        dx, dy = dxdy[direction]
        newX, newY = x + dx, y + dy

    map[y][x], map[newY][newX] = 'X' if isPart1 else map[newY][newX], map[y][x]
    return (False, newX, newY, direction)

# Part 1
x, y = locateRobot(lab)
newLab = [row[:] for row in lab]
direction = directions[0]
leftArea = False
while not leftArea:
    leftArea, x, y, direction = checkObstructions(newLab, x, y, direction, True)

uniqueLocations = 0
for row in newLab:
    uniqueLocations += row.count('X')
# checkLab(newLab)
print('Part 1:', uniqueLocations)

# Part 2
loops = 0
for i in range(len(lab)):
    for j in range(len(lab[i])):
        # print(i, j)
        statesSet = set()
        newLab = [row[:] for row in lab]
        newLab[i][j] = '#'
        x, y = locateRobot(lab)
        direction = directions[0]
        
        leftArea, x, y, direction = checkObstructions(newLab, x, y, direction, False)
        while not leftArea:
            if (x, y, direction) in statesSet:
                loops += 1
                break
            statesSet.add((x, y, direction))
            leftArea, x, y, direction = checkObstructions(newLab, x, y, direction, False)
print('Part 2:', loops)
